src/Google.Bigquery.V2/BigqueryClientImpl.InsertData.cs

```diff
@@ -0,0 +1,165 @@
+        /// <inheritdoc />
+        public override Job UploadJson(TableReference tableReference, TableSchema schema, Stream input)
+        {
+            Preconditions.CheckNotNull(tableReference, nameof(tableReference));
+            Preconditions.CheckNotNull(input, nameof(input));
+            if (schema == null)
+            {
+                // TODO: Handle "table doesn't exist" more reasonably.
+                var table = GetTable(tableReference);
+                schema = table.Schema;
+            }
+
+            var job = new Job
+            {
+                Configuration = new JobConfiguration
+                {
+                    Load = new JobConfigurationLoad
+                    {
+                        DestinationTable = tableReference,
+                        SourceFormat = "NEWLINE_DELIMITED_JSON",
+                        Schema = schema
+                    },
+                },
+            };
+
+            var mediaUpload = Service.Jobs.Insert(job, ProjectId, input, "text/json");
+
+            var finalProgress = mediaUpload.Upload();
+            if (finalProgress.Exception != null)
+            {
+                Console.WriteLine(finalProgress.Exception);
+                throw finalProgress.Exception;
+            }
+            return mediaUpload.ResponseBody;
+        }
+
+        /// <inheritdoc />
+        public override void InsertRow(TableReference tableReference, IDictionary<string, object> rowData, string insertId = null)
+        {
+            Preconditions.CheckNotNull(tableReference, nameof(tableReference));
+            Preconditions.CheckNotNull(rowData, nameof(rowData));
+
+            var body = new TableDataInsertAllRequest
+            {
+                Rows = new[] { new TableDataInsertAllRequest.RowsData { InsertId = insertId, Json = rowData } }
+            };
+            var response = Service.Tabledata.InsertAll(body, tableReference.ProjectId, tableReference.DatasetId, tableReference.TableId).Execute();
+            if (response.InsertErrors != null)
+            {
+                throw new Exception("Error inserting data");
+            }
+        }
+
+        // TODO: Insert IDs?
+        /// <inheritdoc />
+        public override void InsertRows(TableReference tableReference, IEnumerable<IDictionary<string, object>> rowData)
+        {
+            Preconditions.CheckNotNull(tableReference, nameof(tableReference));
+            Preconditions.CheckNotNull(rowData, nameof(rowData));
+
+            var body = new TableDataInsertAllRequest
+            {
+                Rows = rowData.Select(row => new TableDataInsertAllRequest.RowsData { Json = ConvertRowData(row) }).ToList()
+            };
+            var response = Service.Tabledata.InsertAll(body, tableReference.ProjectId, tableReference.DatasetId, tableReference.TableId).Execute();
+            if (response.InsertErrors != null)
+            {
+                throw new Exception("Error inserting data");
```

Comment:
Done (and refactored). Reusing GoogleApiException as that's the exception other things will throw.